src/librustc_typeck/check/demand.rs

```diff
@@ -239,6 +239,21 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (expected_ty, &ty::TyRef(_, checked_ty)) => {
+                if expected_ty == &checked_ty.ty.sty {
+                    let sp = self.sess().codemap().call_span_if_macro(expr.span);
+                    if let Ok(src) = self.tcx.sess.codemap().span_to_snippet(sp) {
```

Comment:
I think processing the snippets is really fragile. Imo it would be much cleaner to destructure `expr` and check for [`ExprAddrOf`](http://manishearth.github.io/rust-internals-docs/rustc/hir/enum.Expr_.html#variant.ExprAddrOf)